src/signature/mod.rs

```diff
@@ -72,19 +126,94 @@ mod test {
     use super::*;
     // The ed25519 and schnorrkel libs use different implementations of `OsRng`
     // two different libraries are used: `rand` and `rand_core` as a workaround
-    use ed25519_dalek::Keypair as edKeypair;
     use rand::prelude::*;
     use rand_core::OsRng;
-    use schnorrkel::Keypair as srKeypair;
 
-    fn generate_ed25519_keypair() -> edKeypair {
+    use ed25519_dalek::{
+        Keypair as Ed25519Keypair, KEYPAIR_LENGTH as ED25519_KEYPAIR_LENGTH,
+        SECRET_KEY_LENGTH as ED25519_SECRET_KEY_LENGTH,
+        SIGNATURE_LENGTH as ED25519_SIGNATURE_LENGTH,
+    };
+
+    use schnorrkel::{
+        Keypair as Sr25519Keypair, KEYPAIR_LENGTH as SR25519_KEYPAIR_LENGTH,
+        SECRET_KEY_LENGTH as SR25519_SECRET_KEY_LENGTH,
+        SIGNATURE_LENGTH as SR25519_SIGNATURE_LENGTH,
+    };
+
+    fn generate_random_seed() -> Vec<u8> {
+        (0..32).map(|_| rand::random::<u8>()).collect()
+    }
+
+    /// copied from wasm/wasm-crypto
+    fn create_from_seed(seed: &[u8]) -> Ed25519Keypair {
+        let seckey = ed25519_dalek::SecretKey::from_bytes(seed).unwrap();
+        let pubkey: Ed25519PublicKey = (&seckey).into();
+
+        create_ed25519_keypair_from_keys(pubkey.as_bytes(), seed).unwrap()
+    }
+
+    /// copied from wasm/wasm-crypto
+    fn create_ed25519_keypair_from_seed(seed: &[u8]) -> Vec<u8> {
+        create_from_seed(seed).to_bytes().to_vec()
+    }
+
+    fn create_sr25519_keypair_from_seed(seed: &[u8]) -> Vec<u8> {
```

Comment:
This is where to check the key format errors
`schnorrkel::MiniSecretKey::from_bytes(seed).unwrap()` will fail if the format is bad.
can return the error here
also not good to use `unwrap` unless you know it will succeed, otherwise use `expect("some message")` if you really want to panic. Best thing is to return the error in libraries